### ISSUE TITLE
Work with a 100GB file is dozens of times slower than with a 4GB file

### ISSUE THREAD

**Issue by Hi-Angel** [openzfs/zfs#10231]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Ubuntu
Distribution Version	| 18.04
Linux Kernel	| 5.4.28-050428-generic
Architecture	| x86_64
ZFS Version	| 0.8.3
SPL Version	| 0.8.3

### Describe the problem you're observing

When a file-system is created over ZFS-provided block device, it is found that work with a bigger file is a lot slower than with a smaller one. FSes were tested are NTFS and EXT4. Workload it's tested on: 70% reads, 30% writes, random, 20 minutes long. Difference is most noticeable after the 2nd test.

It is believed not to be related to caches of any kind, because:

1. target file is opened with `o_direct` option to bypass FS-cache
2. amount of data both written and read in case with smaller file is a dozen times the amount in larger file case. This implies the problem is not because of possibly a bigger number of writes to ZIL *(i.e. because in that case amount of data written should be larger for bigger file instead)*

Amounts of writes and reads I mentioned are according to diskstats. Below is the table. In this specific case, there were separate slog and l2arc devices to have more detailed statistics. However this is reproducible without having them separately as well.

<details>
<summary>Diskstats statistics table</summary>

Description | 100G                             | 4G
------:|:-------------------------------:|:----
**Test1**:
*SLOG*:
Written |7230.40625 MB           | 2349.2734375 MB
Time spent writing |390.982 sec  |152.669 sec
Read |0.0 MB                     | 0.0 MB
Time spent reading | 0.0 sec      |0.0 sec
*L2ARC*:
Written |53066.234375 MB         | 77198.1328125 MB
Time spent writing |8352.832 sec |13555.253 sec
Read |0.0 MB                     | 0.0 MB
Time spent reading | 0.0 sec      |0.0 sec
*zd0*:
Written |42194.3671875 MB        | 96414.6953125 MB
Time spent writing |792.734 sec  |370.697 sec
Read |14966.9765625 MB           | 225008.8984375 MB
Time spent reading | 151.422 sec  |684.358 sec
**Test2**:
*SLOG*:
Written |1516.09375 MB           | 2361.9375 MB
Time spent writing |96.29 sec    |166.099 sec
Read |0.0 MB                     | 0.0 MB
Time spent reading | 0.0 sec      |0.0 sec
*L2ARC*:
Written |7016.1875 MB            | 77203.3359375 MB
Time spent writing |990.418 sec  |13035.121 sec
Read |0.0 MB                     | 0.0 MB
Time spent reading | 0.0 sec      |0.0 sec
*zd0*:
Written |3892.42578125 MB        | 96491.53515625 MB
Time spent writing |100.124 sec  |372.696 sec
Read |1602.921875 MB             | 225186.765625 MB
Time spent reading | 17.948 sec   |681.586 sec
**Test3**:
*SLOG*:
Written |1610.49609375 MB        | 2397.07421875 MB
Time spent writing |100.98 sec   |152.583 sec
Read |0.0 MB                     | 0.0 MB
Time spent reading | 0.0 sec      |0.0 sec
*L2ARC*:
Written |7153.3359375 MB         | 77220.30859375 MB
Time spent writing |1008.907 sec |12755.576 sec
Read |0.0 MB                     | 0.0 MB
Time spent reading | 0.0 sec      |0.0 sec
*zd0*:
Written |4141.53125 MB           | 96504.26171875 MB
Time spent writing |113.123 sec  |370.559 sec
Read |1804.4296875 MB            | 225215.3359375 MB
Time spent reading | 19.665 sec   |691.057 sec
</details>

When this happens, CPUs are almost idle, so I assume there's some locking/sleeping problem.

In "steps to reproduce" I'm using vdbench because it's the main testing utility for us, however the problem was reproduced with `fio` as well.

Any hints on what to look at while debugging this problem are welcome.

### Steps to reproduce

```bash
$ zpool create pool-raidz1-device raidz1 /dev/sd{g,h,i,j,k,m,o,p,r,v}1
$ zfs create -V 500G -b 8K pool-raidz1-device/disk1
$ mkfs.ext4 /dev/zd0 # that is assuming zd0 is the newly created device
$ mount /dev/zd0 /mnt/test
$ truncate -s 100G /mnt/test/file1
$ cat vdbench/zd0.cfg
report_run_totals=yes
sd=sd1,lun=/mnt/test/file1,openflags=o_direct,threads=1,size=4G
wd=wd1,sd=sd*,seekpct=random,xfersize=8k,rdpct=70
rd=rd1,wd=wd1,iorate=max,elapsed=20m,interval=5,warmup=0,threads=1
$ cd vdbench
$ for i in {1..3}; do ./vdbench -f zd0.cfg -o 4G_test$i; done
```

Wait for the 3 tests to finish. Note this was a 4G test. Now:

```bash
$ sed -i 's/4G/100G/' zd0.cfg
$ for i in {1..3}; do ./vdbench -f zd0.cfg -o 100G_test$i; done
```

Compare IOPS number for 100G and 4G cases.

#### Expected

They're more-or-less equal

#### Actual

The difference is a dozen of times since the 2nd test:

Test# | 4G, IOPS | 100G, IOPS
-----:|:---:|:---
Test1 | 25854.5 | 20800.8
Test2 | 33491.4 | 1270.1
Test3 | 33328.6 | 1395.3

**Comment by Hi-Angel** [comment]
Ok, so, what I tried today so far:

1. Measuring a `zio_delay_interrupt()` that I spotted in one of `perf` reports, but it sleeps just around 1ms *(usually less)* per second, so not the culprit. I shall mention though, [perf doesn't seem to allow measuring latency ATM](https://bugzilla.kernel.org/show_bug.cgi?id=207323), so not useful here anyway.
2. Getting ZFS debug prints by executing `echo 1 > /sys/module/zfs/parameters/zfs_flags`. I haven't got any new prints in `journalctl` or `dmesg` or `/proc/spl/kstat/zfs/dbgmsg` though.
3. Comparing the bunch of similar lines that `/proc/spl/kstat/zfs/dbgmsg` has by default in 4GB and 100GB cases. Haven't spotted anything suspicious.

As I'm completely unfamiliar with the codebase, any hints on what to look at are welcome.

**Comment by rincebrain** [comment]
You need to also set zfs_dbgmsg_enable to get anything useful out of dbgmsg, I believe.

**Comment by Hi-Angel** [comment]
> You need to also set zfs_dbgmsg_enable to get anything useful out of dbgmsg, I believe.

Thanks, it turned out I already had it set by the time of writing the comment.

------

Okay, so I just found an interesting behavior. After testing for a while IOPS for some reason got improved for me, so I decided to re-create the pool per steps to reproduce, except that this time I went straight to testing the 100GB case. And I still got somewhat good results ≈20k IOPS. After letting it run for a few minutes I interrupted it and then I tried 4GB. And I got ≈4k IOPS!

Then I recreated everything again, and tried 4GB for a few minutes first and then 100GB second. This time it was reproduced as per the report: i.e. first 100GB test *(20 minutes long)* has shown ≈20k IOPS, and then after the second it got quickly dropped down to 2k and below.

------

I made a few system-wide latency records, the SVG files are attached below. Duration of all records was 30 seconds. *(I had to put them to an archive because github doesn't accept SVG files)*. I used the [offwaktime script](https://github.com/iovisor/bcc/blob/master/tools/offwaketime_example.txt) with a flamegraph. The results are read as follows: there are two stacks separated by `--`. The bottom part is the latency, while the top part are stacks of whoever was waking the thread up. Measurements are in μs. If anyone wants to reproduce, note that using systemd-wide profile with flamegraphs [requires removing idle threads from the output first](https://github.com/jonhoo/inferno/issues/170#issuecomment-617155493).

* 4GB, high performance: [4G-offwaketime.svg.gz](https://github.com/openzfs/zfs/files/4515720/4G-offwaketime.svg.gz)
* 100GB, somewhat low performance, record 1: [100G-offwaketime.svg.gz](https://github.com/openzfs/zfs/files/4515723/100G-offwaketime.svg.gz)
* 100GB, lowest performance, record 2, including userspace stacks *(there's none though, I mean I included that in options)*:  [out.svg.gz](https://github.com/openzfs/zfs/files/4515734/out.svg.gz)
* 100GB, lowest performance, record 3: [out2.svg.gz](https://github.com/openzfs/zfs/files/4515739/out2.svg.gz)

Screenshot versions of some:

4GB-high-performance:
![Screenshot_20200422_143026](https://user-images.githubusercontent.com/7935057/79976764-d2d70a80-84a5-11ea-863d-69ac77cb9d4a.png)

100GB, record 2:
![Screenshot_20200422_143241](https://user-images.githubusercontent.com/7935057/79976950-28abb280-84a6-11ea-8eec-69552901c0f4.png)

I see that in cases with low performance `z_wr_int` takes the most time.

**Comment by adamdmoss** [comment]
One note:

> It is believed not to be related to caches of any kind, because:
>     target file is opened with o_direct option to bypass FS-cache

ZFS currently ignores o_direct I believe - its own (ARC) caching is not bypassed.

**Comment by Hi-Angel** [comment]
> ZFS currently ignores o_direct I believe - its own (ARC) caching is not bypassed.

Thanks, good point.

----

Okay, so to give some update: it is possible that the reason lies outside of ZFS. I measured time between calls to `zvol_request` which does reads and writes, and it turned out to be noticeably larger in case of low performance. For low performance it varies around 18μs..7ms, whereas for good performance it is mostly in xx μs, in rare cases it goes up to 2хх μs.

This discovery made me try testing a similar setup but with LVM + ext4 intead, and I found the same pattern. So at this point it doesn't look specific to ZFS.

Then I tried LVM + xfs, and… well, there's a different pattern though. So far 100GB has always been faster than 4GB for 70% reads/30% writes profile, although it is slower in 100% reads and 100% writes.

Either way, I'm continue researching on this, I hope I'll come with more specific conclusions.

**Comment by Hi-Angel** [comment]
So, I conducted some experiments, and found out that this situation is specific to file systems rather than to ZFS as a volume manager, LVM, etc. I can even reproduce it with EXT4 over a plain SSD.

Another FS I tested was XFS, and the situation there is quite different. XFS works faster with bigger file on "70% reads/30% writes, random" profile compared to XFS for smaller file. At the same time it is somewhat slower with bigger files compared to its work with smaller files when a "100% random writes" or "100% random reads" profile is used. Still, the difference is not that big as it is seen on EXT4.

Tests with EXT4 and XFS over a raw SSD device were conducted on 5.6.2 kernel.

Out of curiosity I also tested ZFS as just a file-system *(i.e. rather than using it as a volume manager like in steps-to-reproduce)*, and am seeing same drop in performance.

Given the original problem I worked on was seen on NTFS *(Windows driver)* over a ZFS-provided block device, and that NTFS driver has no source code, there's nothing we can do about it. Regarding EXT4 I created a report though https://www.spinics.net/lists/linux-ext4/msg71862.html but it's not something that my employer feels worth investing time at. Same regarding ZFS used as a file-system.

I'm closing this report since it turned out that whatever the steps-to-reproduce leads to can not be solved inside ZFS.
